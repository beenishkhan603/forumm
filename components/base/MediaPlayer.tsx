import { playOnElement } from '@libs/Agora/util'
import {
  ILocalAudioTrack,
  ILocalVideoTrack,
  IRemoteAudioTrack,
  IRemoteVideoTrack,
} from 'agora-rtc-sdk-ng'
import React, { useEffect, useRef } from 'react'
import Box from './Box'

export interface VideoPlayerProps {
  videoTrack: ILocalVideoTrack | IRemoteVideoTrack | undefined
  audioTrack: ILocalAudioTrack | IRemoteAudioTrack | undefined
  key: string
  fitType?: 'contain' | 'cover' | 'fill'
  className?: string
  children?: React.ReactNode
  mirror?: boolean
  style?: Object
  muteAudio?: boolean
}

const MediaPlayer = ({
  key,
  videoTrack,
  audioTrack,
  children,
  className = 'relative',
  fitType = 'contain',
  mirror = false,
  style = {},
  muteAudio = false,
}: VideoPlayerProps) => {
  const container = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!container.current) return
    if (!!videoTrack?.play)
      return videoTrack?.play(container.current, {
        fit: fitType,
        mirror: mirror,
      })

    const teardown = playOnElement(
      videoTrack as unknown as MediaStream,
      container.current
    )

    return () => {
      if (videoTrack?.stop) videoTrack?.stop()
      if (teardown) teardown()
    }
  }, [container, fitType, mirror, videoTrack])

  useEffect(() => {
    if (audioTrack) {
      if (!muteAudio) {
        if (!!audioTrack.play) audioTrack.play()
      } else {
        if (!!audioTrack.stop) audioTrack.stop()
      }
    }
    return () => {
      if (!!audioTrack?.stop) audioTrack?.stop()
    }
  }, [audioTrack, muteAudio])

  return (
    <Box
      key={key}
      innerRef={container}
      className={`${className}`}
      style={style}
    >
      {children}
    </Box>
  )
}

export default MediaPlayer
