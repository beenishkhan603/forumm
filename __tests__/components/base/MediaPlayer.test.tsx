import MediaPlayer from '@components/base/MediaPlayer'
import { render, waitFor, screen } from '@testing-library/react'
import { ILocalAudioTrack, ILocalVideoTrack } from 'agora-rtc-sdk-ng'

describe('MediaPlayer component test suite', () => {
  it('SHOULD match the snapshot', () => {
    const { container } = render(
      <MediaPlayer videoTrack={undefined} audioTrack={undefined} />
    )
    expect(container).toMatchSnapshot()
  })

  it('SHOULD play audio on load', () => {
    const audioMock = jest.fn()
    render(
      <MediaPlayer
        videoTrack={undefined}
        audioTrack={
          { play: audioMock, stop: jest.fn() } as unknown as ILocalAudioTrack
        }
      />
    )
    waitFor(async () => {
      expect(audioMock).toHaveBeenCalled()
    })
  })

  it('SHOULD start video WHEN component is mounted', () => {
    jest.mock('react', () => {
      const realModule = jest.requireActual('react')
      return {
        ...realModule,
        useRef() {
          return {
            current: true,
          }
        },
      }
    })
    const mediaMock = jest.fn()
    render(
      <MediaPlayer
        audioTrack={
          { play: jest.fn(), stop: jest.fn() } as unknown as ILocalAudioTrack
        }
        videoTrack={
          { play: mediaMock, stop: jest.fn() } as unknown as ILocalVideoTrack
        }
      />
    )
    waitFor(async () => {
      expect(mediaMock).toHaveBeenCalled()
    })
  })

  it('SHOULD not start video WHEN component isnt mounted', () => {
    jest.mock('react', () => {
      const realModule = jest.requireActual('react')
      return {
        ...realModule,
        useRef() {
          return {
            current: false,
          }
        },
      }
    })
    const mediaMock = jest.fn()
    render(
      <MediaPlayer
        audioTrack={
          { play: jest.fn(), stop: jest.fn() } as unknown as ILocalAudioTrack
        }
        videoTrack={
          { play: mediaMock, stop: jest.fn() } as unknown as ILocalVideoTrack
        }
      />
    )
    waitFor(async () => {
      expect(mediaMock).not.toHaveBeenCalled()
    })
  })
})
