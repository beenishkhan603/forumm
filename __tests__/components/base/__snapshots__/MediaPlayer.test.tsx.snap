// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaPlayer component test suite SHOULD match the snapshot 1`] = `
<div>
  <div
    class="relative"
  />
</div>
`;
